### PR TITLE
Fix some warnings in orttraining code

### PR DESCRIPTION
**Pull request by snnn**
**Description**: 

1. Fix some warnings in orttraining code
2. Use constexpr to replace const when applicable.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/core/framework/pipeline.cc [hunk 1]
```diff
@@ -298,7 +301,7 @@ void PipelineScheduler::CreateFullSchedule() {
       }
 
       // Get the only action in this slot.
-      const auto action = slot.GetFrontAction();
+      const auto& action = slot.GetFrontAction();
```
Comment (snnn) [inline orttraining/orttraining/core/framework/pipeline.cc:304]: Eliminate  unnecessary copy.

File: orttraining/orttraining/core/framework/pipeline.h [hunk 2]
```diff
@@ -125,9 +125,9 @@ class PipelineSlot {
   // Print this structure following a fixed-length format.
   // It assumes there are at most 2 actions per slot.
   friend std::ostream& operator<<(std::ostream& stream, const PipelineSlot& slot);
-  void SetWaitedEvent(const std::vector<int> event);
+  void SetWaitedEvent(const std::vector<int>& event);
```
Comment (snnn) [inline orttraining/orttraining/core/framework/pipeline.h:128]: It should either be "const std::vector<int>& event" or "std::vector<int> event".  A const object shouldn't be moved.

File: orttraining/orttraining/core/framework/pipeline.h [hunk 3]
```diff
@@ -322,13 +322,13 @@ struct PipelineTensorNames {
 
 struct PipelineContext {
   // Number of pipeline stages.
-  int num_pipeline_stages;
+  int num_pipeline_stages = 0;
```
Comment (snnn) [inline orttraining/orttraining/core/framework/pipeline.h:325]: Give every variable a default value to make VC static analyzer happy. Also, I feel it's safer.

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 4]
```diff
@@ -623,7 +623,7 @@ IMPLEMENT_GRADIENT_BUILDER(GetTransposeGradient) {
   if (attributes.empty()) {
     const TensorShapeProto& input_shape = I(0).type_proto->tensor_type().shape();
     if (input_shape.dim_size() > 0) {  // input_shape is available
-      int n = input_shape.dim_size() - 1;
+      size_t n = static_cast<size_t>(input_shape.dim_size()) - 1;
```
Comment (snnn) [inline orttraining/orttraining/core/graph/gradient_builder.cc:626]: Use 64-bit arithmetic if the result value will be used as a 64-bit type.

File: orttraining/orttraining/test/graph/bert_toy_fetches.h [hunk 5]
```diff
@@ -4,13 +4,15 @@
 #pragma once
 
 #include <unordered_map>
+#include <array>
```
Comment (snnn) [inline orttraining/orttraining/test/graph/bert_toy_fetches.h:7]: Fix the warning:

orttraining\orttraining\test\graph\bert_toy_fetches.h(13,56): error 6262:  : Function uses '131072' bytes of stack:  exceeds /analyze:stacksize '16384'.. This allocation was for a compiler-generated temporary for 'double [16384]' at line 13.  Consider moving some data to heap.

**Review by SherlockNoMad — APPROVED** [review]
Thanks for fixing these!

**Pull request merged**